README.md

# A minimal PHP implementation of the CrowdTangle API

[![Latest Version on Packagist](https://img.shields.io/packagist/v/chrishardie/crowdtangle-api.svg?style=flat-square)](https://packagist.org/packages/chrishardie/crowdtangle-api)
[![Total Downloads](https://img.shields.io/packagist/dt/chrishardie/crowdtangle-api.svg?style=flat-square)](https://packagist.org/packages/chrishardie/crowdtangle-api)

This is a minimal PHP implementation of the [CrowdTangle API](https://help.crowdtangle.com/en/articles/1189612-crowdtangle-api). It contains a subset of the methods available. I am open to PRs that add extra methods to the client.

Here are a few examples on how you can use the package:

```php
$client = new ChrisHardie\CrowdtangleApi\Client($accessToken);

// get lists
$client->getLists();

// get accounts in a list
$client->getAccountsForList($listId);

// get posts
$client->getPosts([
    'accounts' => '12345678',
    'startDate' => '2022-03-01',
]);

// get a single post
$client->getPost($postId);
```

View the [full CrowdTangle API Documentation](https://github.com/CrowdTangle/API/wiki) for details on available parameters and syntax.

## Installation

You can install the package via composer:

```bash
composer require chrishardie/crowdtangle-api
```

## Usage

Here are the API methods currently supported:

* `getLists()` - Retrieve the lists, saved searches and saved post lists for an account
* `getAccountsForList($listId, $parameters, $maxRecords)` - Retrieve the accounts for a given list
* `getPosts($parameters, $maxRecords)` - Retrieve a set of posts for the given parameters
* `getPost($postId)` - Retrieves a specific post

In most cases the library is simply passing the required arguments to the CrowdTangle API.

In the case of methods that support pagination (currently, `getAccountsForList()` and  `getPosts()`), by default this library will attempt to retrieve all records on all pages, 100 at a time up to 1000 maximum. You can change this by passing a lower value for `$maxRecords`. Note that CrowdTangle API throttling limits may apply.

## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Credits

- [Chris Hardie](https://github.com/ChrisHardie)
- [All Contributors](../../contributors)

Inspired and structured after Spatie's [Dropbox API](https://github.com/spatie/dropbox-api).

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `crowdtangle-api` will be documented in this file.

## 1.0.0 - 2022-04-04

- initial release


src/Exceptions/BadRequest.php
<?php

namespace ChrisHardie\CrowdtangleApi\Exceptions;

use Exception;
use JsonException;
use Psr\Http\Message\ResponseInterface;

class BadRequest extends Exception
{
    /**
     * @var ResponseInterface
     */
    public ResponseInterface $response;

    /**
     * The CrowdTangle error code supplied in the response.
     *
     * @var int|null
     */
    public ?int $crowdtangleCode;

    /**
     * @throws JsonException
     */
    public function __construct(ResponseInterface $response)
    {
        $this->response = $response;

        $body = json_decode($response->getBody(), true, 512, JSON_THROW_ON_ERROR);

        if ($body !== null) {
            if (isset($body['code'])) {
                $this->crowdtangleCode = $body['code'];
            }

            parent::__construct($body['message']);
        }
    }
}


src/Client.php
<?php

namespace ChrisHardie\CrowdtangleApi;

use ChrisHardie\CrowdtangleApi\Exceptions\BadRequest;
use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\ClientInterface;
use GuzzleHttp\Exception\ClientException;
use GuzzleHttp\Exception\GuzzleException;
use JsonException;

class Client
{
    /**
     * @var TokenProvider
     */
    private TokenProvider $tokenProvider;

    /**
     * @var GuzzleClient
     */
    protected GuzzleClient $client;

    /**
     * @param string|null $accessToken
     * @param GuzzleClient|null $client
     */
    public function __construct(string $accessToken = null, ClientInterface $client = null)
    {
        if ($accessToken instanceof TokenProvider) {
            $this->tokenProvider = $accessToken;
        }
        if (is_string($accessToken)) {
            $this->tokenProvider = new InMemoryTokenProvider($accessToken);
        }

        $this->client = $client ?? new GuzzleClient();
    }

    /**
     * Retrieve the lists, saved searches and saved post lists
     *
     * @link https://github.com/CrowdTangle/API/wiki/lists
     *
     * @return array
     * @throws BadRequest
     * @throws GuzzleException
     * @throws JsonException
     */
    public function getLists(): array
    {
        $body = $this->endpointRequest('lists');
        return $body['result']['lists'];
    }

    /**
     * Retrieve the accounts for a given list.
     * Accounts may only be retrieved for lists of type LIST, as saved searches and saved posts
     * do not have associated accounts.
     *
     * @link https://github.com/CrowdTangle/API/wiki/List-Accounts
     *
     * @param int        $listId
     * @param array|null $parameters
     * @param int        $max_records   The maximum number of records to retrieve across pagination calls
     * @return array
     * @throws BadRequest
     * @throws GuzzleException
     * @throws JsonException
     */
    public function getAccountsForList(int $listId, ?array $parameters = [], int $max_records = 1000): array
    {
        // The API supports a maximum of 100 records at a time
        $parameters['count'] = ! empty($parameters['count'])
            && is_numeric($parameters['count'])
            && $parameters['count'] <= 100
            ? $parameters['count']
            : min(100, $max_records);

        $body = $this->endpointRequest('lists/' . $listId . '/accounts', $parameters);
        $accounts = $body['result']['accounts'];

        // If additional accounts under the maximum are available, get them.
        $loop = 1;
        while (count($accounts) < $max_records && ! empty($body['result']['pagination']['nextPage'])) {
            $parameters['offset'] = $parameters['count'] * $loop;
            $body = $this->endpointRequest('lists/' . $listId . '/accounts', $parameters);
            foreach ($body['result']['accounts'] as $account) {
                $accounts[] = $account;
            }
            $loop++;
        }

        return $accounts;
    }

    /**
     * Retrieve a set of posts for the given parameters.
     *
     * @link https://github.com/CrowdTangle/API/wiki/Posts
     *
     * @param array|null $parameters
     * @param int        $max_records
     * @return array
     * @throws BadRequest
     * @throws GuzzleException
     * @throws JsonException
     */
    public function getPosts(?array $parameters = [], int $max_records = 1000): array
    {
        // The API supports a maximum of 100 records at a time
        $parameters['count'] = ! empty($parameters['count'])
        && is_numeric($parameters['count'])
        && $parameters['count'] <= 100
            ? $parameters['count']
            : min(100, $max_records);

        $body = $this->endpointRequest('posts', $parameters);
        $posts = $body['result']['posts'];

        // If additional posts under the maximum are available, get them.
        $loop = 1;
        while (count($posts) < $max_records && ! empty($body['result']['pagination']['nextPage'])) {
            $parameters['offset'] = $parameters['count'] * $loop;
            $body = $this->endpointRequest('posts', $parameters);
            foreach ($body['result']['posts'] as $post) {
                $posts[] = $post;
            }
            $loop++;
        }

        return $posts;
    }

    /**
     * Retrieves a specific post.
     * The ID format for Facebook and Instagram are different.
     * For Instagram, it's [post_id]_[page_id]
     * For Facebook, it's [page_id]_[post_id]
     *
     * @link https://github.com/CrowdTangle/API/wiki/Posts#get-postid
     *
     * @param string $postId
     * @return array
     * @throws BadRequest
     * @throws GuzzleException
     * @throws JsonException
     */
    public function getPost(string $postId): array
    {
        $body = $this->endpointRequest('post/' . $postId);
        if (! empty($body['result']['posts'][0])) {
            return $body['result']['posts'][0];
        }
        return [];
    }

    /**
     * @param string     $endpoint
     * @param array|null $arguments
     * @return array
     *
     * @throws BadRequest
     * @throws GuzzleException
     * @throws JsonException
     */
    public function endpointRequest(string $endpoint, ? array $arguments = []): array
    {
        try {
            $response = $this->client->get($this->getEndpointUrl($endpoint), [
                'headers' => $this->getHeaders(),
                'query' => $arguments,
            ]);
        } catch (ClientException $exception) {
            $response = $this->endpointRequest($endpoint, $arguments);
        }

        return json_decode($response->getBody(), true, 512, JSON_THROW_ON_ERROR) ?? [];
    }

    protected function getEndpointUrl(string $endpoint): string
    {
        return "https://api.crowdtangle.com/{$endpoint}";
    }

    /**
     * Get the access token.
     */
    public function getAccessToken(): string
    {
        return $this->tokenProvider->getToken();
    }

    /**
     * Set the access token.
     */
    public function setAccessToken(string $accessToken): self
    {
        $this->tokenProvider = new InMemoryTokenProvider($accessToken);

        return $this;
    }

    /**
     * Get the HTTP headers.
     */
    protected function getHeaders(array $headers = []): array
    {
        $auth = [];
        if ($this->tokenProvider) {
            $auth = $this->getHeadersForApiToken($this->tokenProvider->getToken());
        }

        return array_merge($auth, $headers);
    }

    /**
     * @param $token
     * @return array
     */
    protected function getHeadersForApiToken($token): array
    {
        return [
            'x-api-token' => $token,
        ];
    }

    /**
     * @throws JsonException
     */
    protected function determineException(ClientException $exception): BadRequest|ClientException
    {
        if (in_array($exception->getResponse()->getStatusCode(), [400, 409])) {
            return new BadRequest($exception->getResponse());
        }

        return $exception;
    }
}
